controllers/front/OrderConfirmationController.php

```diff
@@ -45,31 +48,92 @@ public function init()
         if (true === (bool) Tools::getValue('free_order')) {
             $this->checkFreeOrder();
         }
 
-        $this->id_cart = (int) (Tools::getValue('id_cart', 0));
-
-        $redirectLink = 'index.php?controller=history';
-
+        // We get the required data from URL
         $this->id_module = (int) (Tools::getValue('id_module', 0));
-        $this->id_order = Order::getIdByCartId((int) ($this->id_cart));
+        $this->id_order = (int) (Tools::getValue('id_order', 0));
         $this->secure_key = Tools::getValue('key', false);
-        $order = new Order((int) ($this->id_order));
+        $this->order = new Order((int) ($this->id_order));
 
+        // We run all the required checks and if something is smelly, we redirect him to order history
+        // The confirmation link must contain a unique secure key matching the order key,
+        // it prevents user to view other customer's order confirmations
+        $redirectLink = 'index.php?controller=history';
         if (!$this->id_order || !$this->id_module || !$this->secure_key || empty($this->secure_key)) {
             Tools::redirect($redirectLink . (Tools::isSubmit('slowvalidation') ? '&slowvalidation' : ''));
         }
-        $this->reference = $order->reference;
-        if (!Validate::isLoadedObject($order) || $order->id_customer != $this->context->customer->id || $this->secure_key != $order->secure_key) {
+        if (!Validate::isLoadedObject($this->order) || $this->secure_key != $this->order->secure_key) {
             Tools::redirect($redirectLink);
         }
         $module = Module::getInstanceById((int) ($this->id_module));
-        if ($order->module != $module->name) {
+        if ($this->order->module != $module->name) {
             Tools::redirect($redirectLink);
         }
-        $this->order_presenter = new OrderPresenter();
+
+        // If checks passed, initialize customer, we will need him anyway
+        $this->customer = new Customer((int) ($this->order->id_customer));
+    }
+
+    /**
+     * Logic after submitting forms
+     *
+     * @see FrontController::postProcess()
+     */
+    public function postProcess()
+    {
+        // If user filled the password and submitted the form to convert his account on the bottom of screen
+        if (Tools::isSubmit('submitTransformGuestToCustomer')) {
+            // Only variable we need is the password
+            // There is no need to check other variables, because hacker would be kicked out in init(), if he tried to convert another customer
+            $password = Tools::getValue('password');
+
+            // If no password was entered
+            if (empty($password)) {
+                $this->errors[] = $this->trans(
+                    'To convert your account, you must enter a password.',
+                    [],
+                    'Shop.Forms.Help'
+                );
+
+            // If the password is short
```

Comment:
These comments are not needed, `(strlen($password) < Validate::PASSWORD_LENGTH)` is exactly "verify password is long enough"